downstairs/src/extent_inner_raw.rs

```diff
@@ -0,0 +1,2028 @@
+        assert_eq!(inner.sync_index, 1);
+
+        // The context should be written to slot 1
+        inner.write(JobId(12), &[&write], false, IOV_MAX_TEST)?;
+        assert_eq!(inner.sync_index, 1);
+
+        // The context should be written to slot 0, forcing a sync
+        inner.write(JobId(12), &[&write], false, IOV_MAX_TEST)?;
+        assert_eq!(inner.sync_index, 2);
+
+        Ok(())
+    }
+
+    /// If a write successfully put a context into a context slot, but it never
+    /// actually got the data onto the disk, that block should revert back to
+    /// being "unwritten". After all, the data never was truly written.
+    ///
+    /// This test is very similar to test_region_open_removes_partial_writes.
+    #[test]
+    fn test_reopen_marks_blocks_unwritten_if_data_never_hit_disk() -> Result<()>
+    {
+        let dir = tempdir()?;
+        let mut inner =
+            RawInner::create(dir.as_ref(), &new_region_definition(), 0)
+                .unwrap();
+
+        // Partial write, the data never hits disk, but there's a context
+        // in the DB and the dirty flag is set.
+        inner.set_dirty_and_block_context(&DownstairsBlockContext {
+            block_context: BlockContext {
+                encryption_context: None,
+                hash: 1024,
+            },
+            block: 0,
+            on_disk_hash: 65536,
+        })?;
+        drop(inner);
+
+        // Reopen, which should note that the hash doesn't match on-disk values
+        // and decide that block 0 is unwritten.
+        let mut inner =
+            RawInner::create(dir.as_ref(), &new_region_definition(), 0)
+                .unwrap();
+
+        // Writing to block 0 should succeed with only_write_unwritten
+        {
+            let data = Bytes::from(vec![0x66; 512]);
+            let hash = integrity_hash(&[&data[..]]);
+            let block_context = BlockContext {
+                encryption_context: None,
+                hash,
+            };
+            let write = crucible_protocol::Write {
+                eid: 0,
+                offset: Block::new_512(0),
+                data: data.clone(),
+                block_context,
+            };
+            inner.write(JobId(30), &[&write], true, IOV_MAX_TEST)?;
+
+            let mut resp = Vec::new();
+            let read = ReadRequest {
+                eid: 0,
+                offset: Block::new_512(0),
+            };
+            inner.read(JobId(31), &[&read], &mut resp, IOV_MAX_TEST)?;
+
+            // We should get back our data! Block 1 was never written.
+            assert_eq!(
+                resp,
+                vec![ReadResponse {
+                    eid: 0,
+                    offset: Block::new_512(0),
+                    data: BytesMut::from(data.as_ref()),
+                    block_contexts: vec![block_context]
+                }]
+            );
+        }
+
+        Ok(())
+    }
+
+    #[test]
+    fn test_defragment() -> Result<()> {
```

Comment:
I still think there should be more defragment tests, exercising all the possible cases that can be seen.

Reply:
:+1: